### ISSUE TITLE
Recommended dvc practices for evolving input data

### ISSUE THREAD

**Issue by villasv** [iterative/dvc#792]
As part of one of my projects, I have to scrape some data to augment the base dataset. The scraped data changes regularly (weekly) thanks to manual data curation on the original source.

What is the best way to track that data using DVC? I have the following possibilities in mind:

- Simply `dvc add` the data, but following #754, by using copy caches I can simply overwrite the file with the new scraping output. `dvc add` again to update the hashes and caches.
- Run the scraping script with `dvc run`, and when updating is desired I simply reproduce ignoring cache
- Create a dummy dependency file containing the scraping timestamp for example, and put that as a dependency of the `dvc run` scraping command. When updating, I just change the timestamp and reproduce the pipeline
- Lastly, embed the latest run timestamp in the scraping script itself. Less hacky in the DVC side, more hacky in the git side. But I'd have to commit the new `.dvc` files anyway.

Is there another way? Opinions on which is best? Is this a particular use case that could benefit for newer dvc features?

This is not a truly reproducible step, so I believe the first option is the most aligned with the philosophy here... so I won't be disappointed if that's the prevailing opinion.

**Comment by villasv** [comment]
Hmm. I think I should move this question to StackOverflow, we should start using it for this kind of interaction.

**Comment by villasv** [comment]
More or less, because the file in question is being generated inside the project. I’d upload it to S3 only to fetch it with dvc as a dependency.

**Comment by efiop** [comment]
Ah, so you have a second project that generates that file in question on your local machine? If so, support for external local deps is also coming in 0.9.8, so you could use `/path/to/some/latest.data` instead of `s3://mybucket/latest.data` in the commands above. Is that more appropriate for your case?

**Comment by villasv** [comment]
No, I meant that it’s the same project, so it’s a bit weird to push to S3 only to pull it again and cache it back on S3 again with DVC.

But thinking again, it’s not that weird. It’s kinda logical if I want to work with a team to have the most up-to-date input data in a common location. It’s just that I expected DVC to be that location.

FWIW, I’ve been using the copy cache approach and it’s going out fine. The only downside is missing the hardlink optimization, but I think I prefer that over having an extra S3 file upload step.

**Comment by efiop** [comment]
Oh, sorry, I misread this whole thread. Let me try again :)

So you have your project with a base data set, that you need to update(append to) weekly, right? And the problem you are having is that in order to do so you have to copy that file, append to it and then move it in place of the original one, because you don't want to mess up your local cache for that previous version. Correct?

**Comment by villasv** [comment]
Originally, yes. After using the copy cache workaround I don’t have to copy it manually anymore. (And a minor detail is that editing is not append mode, it’s actual updates)

**Comment by efiop** [comment]
Ah, I see. Updating dataset in-place is a very interesting scenario and indeed it doesn't currently work smoothly with dvc. It looks like using `dvc run` to run your scrapper would be the perfect option, however because dvc removes outputs before reproduction it is not possible to implement right now :( 

Maybe we could consider adding some kind of option for the stage file(or the output itself) that will tell dvc to not remove it before reproduction, so that stage could write over the existing output instead of recreating it. It should greatly assist in your scenario without forcing you to do anything by-hand. That being said, on file systems that don't support reflinks, dvc will still have to physically copy your data to unlink it from the cache. What OS/fs combo are you running? If you are on mac with High Sierra or newer, you probably have support for reflinks.

I think we can introduce -p|--protected and -P|--protected-no-cache(along with resulting `protected: True` in respective parts of dvcfiles) options for `dvc run` that will tell dvc to not remove output before reproduction.

What do you think?

**Comment by villasv** [comment]
I think that in the end of the story, I came to the conclusion that this (scripted in-place update) shouldn't be a thing. At least not conceptually visible at the logical level.

What I mean is that if a script updates files in place, there should be some effort by the script itself to behave as an input/output pipeline like any other. That's not hard to achieve: backup the old file, perform the updates on the new one. So even if the `--protected` would solve the problem, it think it would be an anti-pattern. I was doing that by hand because I didn't want to modify the script, but now I think I **should**.

The real issue here to me is not the fact that DVC erases the previous output, it's the one noted by https://github.com/iterative/dvc/issues/754, because by overwriting the previous hardlinked output may corrupt the cache. So the following scenario:

```
python generator.py <-- generates output.txt
dvc add output.txt
...
python generator.py <--- generates a new output.txt
dvc add output.txt <--- will create a new cache entry, but is the previous one corrupted?
```

The only way I could find so far to make this work is to follow the exact same thing as in #754: configure DVC to cache by copy, which is not ideal. Now, I'm wondering if I should also change those scripts to follow the same strategy as I mentioned: write to a new file, overwrite the previous file (therefore erasing the hardlink instead of corrupting the cached data).

I think I might be giving this too much trouble, but after two days of thinking about it and trying out a few solutions, I think the recommended approach should be:

> Don't edit or overwrite added data files in-place.

> If you want to change added data files programmatically, write to a new file and then overwrite the older one in the filesystem. (`mv new old`)

> If you want to curate added data files manually, configure `dvc` to cache them by copy.

**Comment by efiop** [comment]
@villasv Again, so sorry for such a late response.

I agree with your arguments and just wanted to note that the `dvc config cache.type copy` way of doing things although slow, is hands down the most convenient to use. That being said, I want to notice that we also have `reflink` support in dvc, which combines both convenience of `copy` with the speed of hard/symlinks. And if you wish to have those both features at the same time, you could pretty easily transfer your project to a partition that has filesystem that supports `reflinks`. Please correct me if i'm wrong, but I suppose you are probably using Mac as a personal machine that you use to work on a remote Linux server that has your data stored. If it is so, you could consider moving your project to something like XFS partition and have the fastest and the most efficient setup possible from the dvc standpoint :slightly_smiling_face:

Thanks,
Ruslan

**Comment by villasv** [comment]
Pinging back on this issue. What I developed as the routine that would accommodate well with the dvc way of doing things:

1. There's an in-place version of the scraped data, hardlinked to the cache
2. The scraping script will copy the old version to a temporary file and append to it
3. Copying the temp file back, it will overwrite the old version hardlink, being correctly treated as a new file and won't corrupt the cache.

As for hand curation in-place, it's just forbidden. We settled that all curation should be done in the original source, or if very necessary to be an inside venture, versioned in a "patch-like" file that is treated as code instead of data (this versioned with git).

I think I can close the issue. This may not work for everyone, but it has been working for me :)
